client/rest-high-level/src/main/java/org/elasticsearch/client/Request.java

```diff
@@ -68,48 +79,207 @@ static Request get(GetRequest getRequest) {
-        Map<String, String> params = new HashMap<>();
-        putParam("preference", getRequest.preference(), params);
-        putParam("routing", getRequest.routing(), params);
-        putParam("parent", getRequest.parent(), params);
-        if (getRequest.refresh()) {
-            params.put("refresh", Boolean.TRUE.toString());
+        Params parameters = Params.builder();
+        parameters.withPreference(getRequest.preference());
+        parameters.withRouting(getRequest.routing());
+        parameters.withParent(getRequest.parent());
+        parameters.withRefresh(getRequest.refresh());
+        parameters.withRealtime(getRequest.realtime());
+        parameters.withStoredFields(getRequest.storedFields());
+        parameters.withVersion(getRequest.version());
+        parameters.withVersionType(getRequest.versionType());
+        parameters.withFetchSourceContext(getRequest.fetchSourceContext());
+        return parameters.getParams();
+    }
+
+    private static String getEndpoint(GetRequest getRequest) {
+        StringJoiner pathJoiner = new StringJoiner("/", "/", "");
+        return pathJoiner.add(getRequest.index()).add(getRequest.type()).add(getRequest.id()).toString();
+    }
+
+    static Request index(IndexRequest indexRequest) {
+        String method = Strings.hasLength(indexRequest.id()) ? "PUT" : "POST";
+
+        Endpoint endpoint = Endpoint.builder();
+        endpoint.add(indexRequest.index());
+        endpoint.add(indexRequest.type());
+        endpoint.add(indexRequest.id());
+        if (indexRequest.opType() == DocWriteRequest.OpType.CREATE) {
+            endpoint.withSuffix("/_create");
+        }
+
+        Params parameters = Params.builder();
+        parameters.withRouting(indexRequest.routing());
+        parameters.withParent(indexRequest.parent());
+        parameters.withTimeout(indexRequest.timeout());
+        parameters.withVersion(indexRequest.version());
+        parameters.withVersionType(indexRequest.versionType());
+        parameters.withPipeline(indexRequest.getPipeline());
+        parameters.withRefreshPolicy(indexRequest.getRefreshPolicy());
+        parameters.withWaitForActiveShards(indexRequest.waitForActiveShards());
+
+        BytesRef source = indexRequest.source().toBytesRef();
+        ContentType contentType = ContentType.create(indexRequest.getContentType().mediaType());
+        HttpEntity entity = new ByteArrayEntity(source.bytes, source.offset, source.length, contentType);
+
+        return new Request(method, endpoint.getEndpoint(), parameters.getParams(), entity);
+    }
+
+    /**
+     * Utility class to build request's endpoint.
+     */
+    private static class Endpoint {
+        private final List<String> parts = new ArrayList<>();
+        private final String delimiter = "/";
+        private final String prefix = "/";
+        private String suffix = "";
+
+        private Endpoint() {
+        }
+
+        Endpoint withSuffix(String suffix) {
```

Comment:
given that the suffix is also known before we go and provide index, type, and id... (even in case of e.g. _search) I wonder if we can get rid of the list, and just provide parts and suffix as constructor argument, then convert it into string straightaway. Not even sure we need a builder for this.